grid_common/grid_cal.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * grid_cal.c
+ *
+ * Created: 25/11/2024 3:01:37 PM
+ * Author : BENB
+ */
+
+#include "grid_cal.h"
+
+struct grid_cal_model grid_cal_state;
+
+uint8_t grid_cal_init(struct grid_cal_model* cal, uint8_t resolution, uint8_t length) {
+
+  cal->resolution = resolution;
+  cal->length = length;
+
+  cal->maximum = 1 << cal->resolution;
+
+  cal->value = malloc(cal->length * sizeof(uint16_t));
+  cal->center = malloc(cal->length * sizeof(uint16_t));
+  cal->enable = malloc(cal->length * sizeof(uint8_t));
+
+  const uint16_t half_value = cal->maximum >> 1;
+
+  const uint16_t default_offset = +32 * 4.5;
+
+  for (uint8_t i = 0; i < length; ++i) {
+    cal->value[i] = half_value + default_offset;
+    cal->center[i] = cal->value[i];
+    cal->enable[i] = 0;
+  }
+
+  return 0;
+}
+
+uint8_t grid_cal_enable_range(struct grid_cal_model* cal, uint8_t start, uint8_t length) {
+
+  for (uint8_t i = start; i < length; ++i) {
+    cal->enable[i] = 1;
+  }
+
+  return 0;
+}
+
+uint8_t grid_cal_center_set(struct grid_cal_model* cal, uint8_t channel, uint16_t center) {
+
+  if (!cal->enable[channel]) {
+    return 1;
+  }
```

Comment:
only  implement return status if status is actually checked. Otherwise stick to void.
if you decide to implement and check status, make sure it's type is int.